File: fortran/tests/test-radionuclides.f90
PROGRAM test_radionuclides

USE, INTRINSIC :: ISO_C_BINDING
USE, INTRINSIC :: ISO_FORTRAN_ENV
USE :: xraylib
USE :: libtest
IMPLICIT NONE

TYPE(xrl_error), POINTER :: error => NULL()
INTEGER, PARAMETER :: RADIONUCLIDE_LIST_LEN = 10
CHARACTER(KIND=C_CHAR, LEN=RADIO_NUCLIDE_STRING_LENGTH), DIMENSION(:), POINTER :: radionuclide_list
INTEGER :: i
TYPE (radioNuclideData), POINTER :: c

radionuclide_list => GetRadioNuclideDataList(error)
CALL assert(.NOT. ASSOCIATED(error))
CALL assert(SIZE(radionuclide_list) == RADIONUCLIDE_LIST_LEN)

DO i=1,RADIONUCLIDE_LIST_LEN
        c => GetRadioNuclideDataByName(radionuclide_list(i))
        CALL assert(c%name == radionuclide_list(i))
        DEALLOCATE(c)
ENDDO

DO i=1,RADIONUCLIDE_LIST_LEN
        c => GetRadioNuclideDataByIndex(i-1)
        CALL assert(c%name == radionuclide_list(i))
        DEALLOCATE(c)
ENDDO

DEALLOCATE(radionuclide_list)

c => GetRadioNuclideDataByIndex(3)
CALL assert(c%A == 125)
CALL assert(ALL(ABS(c%GammaEnergies - [35.4919_C_DOUBLE]) < 1E-6_C_DOUBLE))
CALL assert(ALL(ABS(c%GammaIntensities - [0.0668_C_DOUBLE]) < 1E-6_C_DOUBLE))
CALL assert(c%N == 72)
CALL assert(ALL(ABS(c%XrayIntensities- [0.0023_C_DOUBLE, 0.00112_C_DOUBLE, &
   0.0063_C_DOUBLE, 0.056_C_DOUBLE, 0.035_C_DOUBLE, 0.0042_C_DOUBLE, &
   0.007_C_DOUBLE, 0.00043_C_DOUBLE, 0.0101_C_DOUBLE, 0.0045_C_DOUBLE, &
   0.00103_C_DOUBLE, 0.0016_C_DOUBLE, 3.24e-05_C_DOUBLE, 0.406_C_DOUBLE, &
   0.757_C_DOUBLE, 0.0683_C_DOUBLE, 0.132_C_DOUBLE, 0.00121_C_DOUBLE, &
   0.0381_C_DOUBLE, 0.0058_C_DOUBLE]) < 1E-6_C_DOUBLE))
CALL assert(ALL(c%XrayLines - [-86, -60, -89, -90, -63, -33, -34, -91, -95,&
   -68, -38, -39, -1, -2, -3, -5, -6, -8, -11, -13] == 0))
CALL assert(c%Z == 53)
CALL assert(c%Z_xray == 52)
CALL assert(c%nGammas == 1)
CALL assert(c%nXrays == 20)
CALL assert(c%name == '125I')
DEALLOCATE(c)

c => GetRadioNuclideDataByName('125I')
CALL assert(c%A == 125)
CALL assert(ALL(ABS(c%GammaEnergies - [35.4919_C_DOUBLE]) < 1E-6_C_DOUBLE))
CALL assert(ALL(ABS(c%GammaIntensities - [0.0668_C_DOUBLE]) < 1E-6_C_DOUBLE))
CALL assert(c%N == 72)
CALL assert(ALL(ABS(c%XrayIntensities- [0.0023_C_DOUBLE, 0.00112_C_DOUBLE, &
   0.0063_C_DOUBLE, 0.056_C_DOUBLE, 0.035_C_DOUBLE, 0.0042_C_DOUBLE, &
   0.007_C_DOUBLE, 0.00043_C_DOUBLE, 0.0101_C_DOUBLE, 0.0045_C_DOUBLE, &
   0.00103_C_DOUBLE, 0.0016_C_DOUBLE, 3.24e-05_C_DOUBLE, 0.406_C_DOUBLE, &
   0.757_C_DOUBLE, 0.0683_C_DOUBLE, 0.132_C_DOUBLE, 0.00121_C_DOUBLE, &
   0.0381_C_DOUBLE, 0.0058_C_DOUBLE]) < 1E-6_C_DOUBLE))
CALL assert(ALL(c%XrayLines - [-86, -60, -89, -90, -63, -33, -34, -91, -95,&
   -68, -38, -39, -1, -2, -3, -5, -6, -8, -11, -13] == 0))
CALL assert(c%Z == 53)
CALL assert(c%Z_xray == 52)
CALL assert(c%nGammas == 1)
CALL assert(c%nXrays == 20)
CALL assert(c%name == '125I')
DEALLOCATE(c)

c => GetRadioNuclideDataByIndex(-1)
CALL assert(.NOT. ASSOCIATED(c))
c => GetRadioNuclideDataByIndex(RADIONUCLIDE_LIST_LEN)
CALL assert(.NOT. ASSOCIATED(c))
c => GetRadioNuclideDataByName('non-existent-radionuclide')
CALL assert(.NOT. ASSOCIATED(c))

c => GetRadioNuclideDataByIndex(-1, error)
CALL assert(.NOT. ASSOCIATED(c))
CALL assert(ASSOCIATED(error))
CALL assert(error%code == XRL_ERROR_INVALID_ARGUMENT)
DEALLOCATE(error)
c => GetRadioNuclideDataByIndex(RADIONUCLIDE_LIST_LEN, error)
CALL assert(.NOT. ASSOCIATED(c))
CALL assert(ASSOCIATED(error))
CALL assert(error%code == XRL_ERROR_INVALID_ARGUMENT)
DEALLOCATE(error)
c => GetRadioNuclideDataByName('non-existent-radionuclide', error)
CALL assert(.NOT. ASSOCIATED(c))
CALL assert(ASSOCIATED(error))
CALL assert(error%code == XRL_ERROR_INVALID_ARGUMENT)
DEALLOCATE(error)

ENDPROGRAM test_radionuclides

File: cplusplus/tests/test-radionuclides.cpp
/* Copyright (c) 2018, Tom Schoonjans
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
    * The names of the contributors may not be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY Tom Schoonjans ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Tom Schoonjans BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

*/

#ifdef NDEBUG
  #undef NDEBUG
#endif
#include "xraylib++.h"
#include "xraylib-error-private.h"
#include <cmath>
#include <cassert>
#include <cstring>

int main(int argc, char *argv[]) {

	std::vector<std::string> nuclides = xrlpp::GetRadioNuclideDataList();
	assert(nuclides.size() == 10);
	
	for (int i = 0 ; i < nuclides.size() ; i++) {
		auto nuclide = nuclides[i];
		auto rnd1 = xrlpp::GetRadioNuclideDataByName(nuclide);
		assert(rnd1.name == nuclide);
		auto rnd2 = xrlpp::GetRadioNuclideDataByIndex(i);
		assert(rnd2.name == nuclide);
	}

	auto rnd = xrlpp::GetRadioNuclideDataByIndex(3);
	assert(rnd.name == "125I");
	assert(rnd.A == 125);
	assert(rnd.N == 72);
	assert(rnd.Z == 53);
	assert(rnd.Z_xray == 52);
	assert(rnd.nGammas== 1);
	assert(fabs(rnd.GammaEnergies[0] - 35.4919) < 1E-4);
	assert(fabs(rnd.GammaIntensities[0] - 0.0668) < 1E-4);
	assert(rnd.nXrays== 20);
	assert(rnd.XrayLines[0] == -86);
	assert(fabs(rnd.XrayIntensities[0] - 0.0023) < 1E-4);

	try {
		xrlpp::GetRadioNuclideDataByIndex(-1);
		abort();
	}
	catch (std::invalid_argument &e) {

	}

	try {
		xrlpp::GetRadioNuclideDataByIndex(10);
		abort();
	}
	catch (std::invalid_argument &e) {

	}

	try {
		xrlpp::GetRadioNuclideDataByName("");
		abort();
	}
	catch (std::invalid_argument &e) {

	}

	try {
		xrlpp::GetRadioNuclideDataByName("jwefhjoehoeoehr");
		abort();
	}
	catch (std::invalid_argument &e) {

	}

	return 0;
}

